x-pack/libbeat/management/managerV2.go

```diff
@@ -0,0 +1,300 @@
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+// BeatV2Manager is the main type for tracing V2-related config updates
+type BeatV2Manager struct {
+	config   *Config
+	registry *reload.Registry
+	client   client.V2
+
+	logger *logp.Logger
+
+	// Track individual units given to us by the V2 API
+	unitsMut sync.Mutex
+	units    map[string]*client.Unit
+	mainUnit string
+
+	// This satisfies the SetPayload() function, and will pass along this value to the UpdateStatus()
+	// call whenever a config is re-registered
+	payload map[string]interface{}
+
+	// stop callback must be registered by libbeat, as with the V1 callback
+	stopFunc func()
+	beatStop sync.Once
+
+	isRunning bool
+}
+
+// NewV2AgentManager returns a remote config manager for the agent V2 protocol.
+// This is meant to be used by the management plugin system, which will register this as a callback.
+func NewV2AgentManager(config *conf.C, registry *reload.Registry, beatUUID uuid.UUID) (lbmanagement.Manager, error) {
+	c := DefaultConfig()
+	if config.Enabled() {
+		if err := config.Unpack(&c); err != nil {
+			return nil, fmt.Errorf("parsing fleet management settings: %w", err)
+		}
+	}
+	agentClient, _, err := client.NewV2FromReader(os.Stdin, client.VersionInfo{Name: "elastic-agent-shipper", Version: "v2"})
+	if err != nil {
+		return nil, fmt.Errorf("error reading control config from agent: %w", err)
+	}
+
+	return NewV2AgentManagerWithClient(c, registry, agentClient)
+}
+
+// NewV2AgentManagerWithClient actually creates the manager instance used by the rest of the beats.
+func NewV2AgentManagerWithClient(config *Config, registry *reload.Registry, agentClient client.V2) (lbmanagement.Manager, error) {
+	log := logp.NewLogger(lbmanagement.DebugK)
+	m := &BeatV2Manager{
+		config:   config,
+		logger:   log.Named("V2-manager"),
+		registry: registry,
+		units:    make(map[string]*client.Unit),
+	}
+
+	if config.Enabled {
+		m.client = agentClient
+	}
+	return m, nil
+}
+
+// ================================
+// Beats central management interface implementation
+// ================================
+
+// UpdateStatus updates the manager with the current status for the beat.
+func (cm *BeatV2Manager) UpdateStatus(status lbmanagement.Status, msg string) {
+	updateState := client.UnitState(status)
+	_ = cm.getUnit(cm.mainUnit).UpdateState(updateState, msg, cm.payload)
+}
+
+// Enabled returns true if config management is enabled.
+func (cm *BeatV2Manager) Enabled() bool {
+	return cm.config.Enabled
+}
+
+// SetStopCallback sets the callback to run when the manager want to shutdown the beats gracefully.
+func (cm *BeatV2Manager) SetStopCallback(stopFunc func()) {
+	cm.stopFunc = stopFunc
+}
+
+// Start the config manager.
+func (cm *BeatV2Manager) Start() error {
+	if !cm.Enabled() {
+		return fmt.Errorf("V2 Manager is disabled")
+	}
+	err := cm.client.Start(context.Background())
+	if err != nil {
+		return fmt.Errorf("error starting connection to client")
+	}
+
+	go cm.unitListen()
```

Comment:
Does this goroutine need to exit cleanly when the beat process itself exits? Is there a done channel or context it can be passed when the beat is signalled to shutdown?

Reply:
Yah, going to refactor this to handle shutdowns in a slightly more centralized way.